File: backend/app/models/purchase.py
import uuid
import enum
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    Column, String, Numeric, Boolean, Date, DateTime, Enum, ForeignKey, CheckConstraint
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.database import Base


class POStatusEnum(enum.Enum):
    draft = "draft"
    confirmed = "confirmed"
    partially_received = "partially_received"
    fully_received = "fully_received"
    cancelled = "cancelled"


class PurchaseOrder(Base):
    __tablename__ = "purchase_orders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    reference = Column(String(40), unique=True, nullable=False)
    vendor_id = Column(
        UUID(as_uuid=True),
        ForeignKey("vendors.id"),
        nullable=False
    )
    vendor_address = Column(String(250), nullable=True)
    responsible_person_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True
    )
    status = Column(
        Enum(POStatusEnum, name="po_status_enum"),
        nullable=False,
        default=POStatusEnum.draft
    )
    expected_delivery_date = Column(Date, nullable=True)
    auto_created = Column(Boolean, nullable=False, default=False)
    source_sales_order_id = Column(
        UUID(as_uuid=True),
        ForeignKey("sales_orders.id"),
        nullable=True
    )
    created_at = Column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    created_by = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True
    )

    # Relationships
    vendor = relationship("Vendor", back_populates="purchase_orders")
    responsible_person = relationship(
        "User",
        foreign_keys=[responsible_person_id],
        back_populates="purchase_orders_as_responsible"
    )
    created_by_user = relationship(
        "User",
        foreign_keys=[created_by],
        back_populates="purchase_orders_created"
    )
    source_sales_order = relationship(
        "SalesOrder",
        back_populates="purchase_orders_sourced"
    )
    lines = relationship("PurchaseOrderLine", back_populates="purchase_order", cascade="all, delete-orphan")


class PurchaseOrderLine(Base):
    __tablename__ = "purchase_order_lines"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    purchase_order_id = Column(
        UUID(as_uuid=True),
        ForeignKey("purchase_orders.id", ondelete="CASCADE"),
        nullable=False
    )
    product_id = Column(
        UUID(as_uuid=True),
        ForeignKey("products.id"),
        nullable=False
    )
    ordered_qty = Column(Numeric(14, 3), nullable=False)
    received_qty = Column(Numeric(14, 3), nullable=False, default=Decimal("0"))
    cost_price = Column(Numeric(14, 2), nullable=False)
    batch_number = Column(String(60), nullable=True)

    __table_args__ = (
        CheckConstraint("ordered_qty > 0", name="purchase_order_lines_ordered_qty_check"),
        CheckConstraint("received_qty >= 0", name="purchase_order_lines_received_qty_check"),
        CheckConstraint("received_qty <= ordered_qty", name="chk_received_not_over"),
    )

    # Relationships
    purchase_order = relationship("PurchaseOrder", back_populates="lines")
    product = relationship("Product", back_populates="purchase_order_lines")

File: frontend/src/types/purchase.ts
export interface PurchaseOrderLine {
  id: string;
  product_id: string;
  product_name: string;
  ordered_qty: number;
  received_qty: number;
  cost_price: number;
  line_total: number;
}

export interface VendorBrief {
  id: string;
  name: string;
  address?: string;
}

export interface UserBrief {
  id: string;
  name: string;
}

export interface PurchaseOrder {
  id: string;
  reference: string;
  vendor: VendorBrief;
  vendor_address?: string;
  responsible_person?: UserBrief;
  status: 'draft' | 'confirmed' | 'partially_received' | 'fully_received' | 'cancelled';
  auto_created: boolean;
  source_sales_order_id?: string;
  source_sales_order_ref?: string;
  lines: PurchaseOrderLine[];
  total_amount: number;
  created_at: string;
  created_by?: UserBrief;
}

export interface PurchaseOrderListItem {
  id: string;
  reference: string;
  vendor_name: string;
  status: string;
  auto_created: boolean;
  total_amount: number;
  created_at: string;
}

export interface PurchaseOrderLineCreate {
  product_id: string;
  ordered_qty: number;
}

export interface PurchaseOrderCreateRequest {
  vendor_id: string;
  responsible_person_id?: string;
  line_items: PurchaseOrderLineCreate[];
}

export interface PurchaseOrderLineReceive {
  line_id: string;
  received_qty: number;
}

export interface PurchaseOrderReceiveRequest {
  lines: PurchaseOrderLineReceive[];
}

export interface Vendor {
  id: string;
  name: string;
}

export interface ProductBrief {
  id: string;
  name: string;
  cost_price: number;
  sales_price: number;
  on_hand_qty: number;
  free_to_use_qty?: number;
}

export const STATUS_COLORS: Record<string, string> = {
  draft: 'bg-slate-500/20 text-slate-300',
  confirmed: 'bg-blue-500/20 text-blue-300',
  partially_received: 'bg-amber-500/20 text-amber-300',
  fully_received: 'bg-emerald-500/20 text-emerald-300',
  cancelled: 'bg-red-500/20 text-red-300',
};

export const STATUS_LABELS: Record<string, string> = {
  draft: 'Draft',
  confirmed: 'Confirmed',
  partially_received: 'Partially Received',
  fully_received: 'Fully Received',
  cancelled: 'Cancelled',
};
